### PR TITLE
Features/issue 15

### PR DESCRIPTION
**Pull request by crankycoder**
This persists events into IndexedDB and schedules a timer interval to upload data using the encrypted ping method in pioneer-utils.  Last upload date is stored in a pref and uploads are restricted from calling the upload method more than once a day.

r? @rehandalal

### CODE REVIEW

File: extension/lib/ActiveURIService.jsm [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 const { interfaces: Ci, utils: Cu } = Components;
 Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
```
Comment (rehandalal) [inline extension/lib/ActiveURIService.jsm]: We don't seem to be using this anywhere in the file?

File: extension/lib/Config.jsm [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+const { utils: Cu } = Components;
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "AddonManager", "resource://gre/modules/AddonManager.jsm");
+
+const EXPORTED_SYMBOLS = ["config"];
+
+const config = {
+  pioneer: {
+    studyName: "pioneer-study-online-news",  // this name must match the pipeline schema to be used
```
Comment (rehandalal) [inline extension/lib/Config.jsm]: the study name is just `"online-news"`

Comment (rehandalal) [reply extension/lib/Config.jsm]: and the comment is not accurate anymore

File: extension/lib/Config.jsm [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    studyVersion: 1,
```
Comment (rehandalal) [inline extension/lib/Config.jsm]: With changes to the pipeline and utils, we don't need this anymore

File: extension/lib/NewsStorage.jsm [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    return allDBPings;
+  },
+
+  async uploadPings() {
+    // upload ping dataset at the most once a day
+    let pioneerUtils = new PioneerUtils(config.pioneer);
+    this.getAllPings().then(payload => {
+      let lastUploadDate = Services.prefs.getCharPref(UPLOAD_DATE_PREF, "");
+      if (lastUploadDate !== isonow()) {
+        pioneerUtils.submitEncryptedPing({entries:payload}).then(() => {
```
Comment (rehandalal) [inline extension/lib/NewsStorage.jsm]: the API for this has changed to accommodate the recent telemetry pipeline changes. it should be:
```js
pioneerUtils.submitEncryptedPing("online-news-log", 1, { entries: payload })
```

File: extension/lib/NewsStorage.jsm [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+          this.clear().then(() => {
+            Services.prefs.setCharPref(UPLOAD_DATE_PREF, isonow());
+          });
+        }, reason => {
+        });
+      } else {
```
Comment (rehandalal) [inline extension/lib/NewsStorage.jsm]: empty else?

File: extension/lib/NewsStorage.jsm [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    // upload ping dataset at the most once a day
+    let pioneerUtils = new PioneerUtils(config.pioneer);
+    this.getAllPings().then(payload => {
+      let lastUploadDate = Services.prefs.getCharPref(UPLOAD_DATE_PREF, "");
+      if (lastUploadDate !== isonow()) {
+        pioneerUtils.submitEncryptedPing({entries:payload}).then(() => {
+          this.clear().then(() => {
+            Services.prefs.setCharPref(UPLOAD_DATE_PREF, isonow());
+          });
+        }, reason => {
```
Comment (rehandalal) [inline extension/lib/NewsStorage.jsm]: do we need this?

File: extension/lib/NewsStorage.jsm [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    // `getStore(db).put` doesn't seem to match up to
+    // https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/put
+    // Invoking a put() always errors out and doesn't call the failure
+    // branch of the promise.  The error in the debugger is:
+    // "DataError: Data provided to an operation does not meet
+    // requirements."
+    // Pings are strictly timestamp ordered.
+    getStore(db).delete(ping.timestamp).then(() => {
+      getStore(db).add(ping).then(() => {
+      }, reason => {
```
Comment (rehandalal) [inline extension/lib/NewsStorage.jsm]: do we need this?

File: extension/lib/NewsStorage.jsm [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+      });
+    }, reason => {
```
Comment (rehandalal) [inline extension/lib/NewsStorage.jsm]: or this?

**Review by rehandalal — CHANGES_REQUESTED** [review]
You'll likely want to upgrade pioneer-utils to v1.0.5 to get the changes for the new telemetry pipeline.

**Comment by crankycoder** [comment]
@rehandalal I've updated the PR as per review comments.

File: extension/lib/Config.jsm [hunk 9]
```diff
@@ -0,0 +1,17 @@
+/* eslint no-unused-vars: ["error", { "varsIgnorePattern": "(config|EXPORTED_SYMBOLS)" }]*/
```
Comment (rehandalal) [inline extension/lib/Config.jsm]: we can actually probably ditch this file and just use the one mythmon created

File: extension/lib/NewsStorage.jsm [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+const { interfaces: Ci, utils: Cu } = Components;
+
+const UPLOAD_DATE_PREF = "pioneer.study.online.news.upload.date";
+
+Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+XPCOMUtils.defineLazyModuleGetter(this,
+  "config",
+  "resource://pioneer-study-online-news/lib/Config.jsm");
```
Comment (rehandalal) [inline extension/lib/NewsStorage.jsm]:
```js
"resource://pioneer-study-online-news/Config.jsm");
```

**Review by rehandalal — CHANGES_REQUESTED** [review]
i landed mythmon's PR which caused some conflicts that need resolution

**Review by rehandalal — APPROVED** [review]
r+. thanks!!

**Pull request merged**